src/url.rs

```diff
@@ -314,6 +354,130 @@ impl PyMultiHostUrl {
+                        ));
+                    }
+                    multi_url.push_str(&single_host.to_string());
+                    if index != hosts.len() - 1 {
+                        multi_url.push(',');
+                    };
+                }
+                multi_url
+            } else if host.is_some() {
+                let url_host = UrlHostParts {
+                    username: username.map(Into::into),
+                    password: password.map(Into::into),
+                    host: host.map(Into::into),
+                    port: port.map(Into::into),
+                };
+                format!("{scheme}://{url_host}")
+            } else {
+                return Err(PyValueError::new_err("expected either `host` or `hosts` to be set"));
+            };
+
+        if let Some(path) = path {
+            url.push('/');
+            url.push_str(path);
+        }
+        if let Some(query) = query {
+            url.push('?');
+            url.push_str(query);
+        }
+        if let Some(fragment) = fragment {
+            url.push('#');
+            url.push_str(fragment);
+        }
+        cls.call1((url,))
+    }
+}
+
+pub struct UrlHostParts {
+    username: Option<String>,
+    password: Option<String>,
+    host: Option<String>,
+    port: Option<u16>,
+}
+
+impl UrlHostParts {
+    fn is_empty(&self) -> bool {
+        self.host.is_none() && self.password.is_none() && self.host.is_none() && self.port.is_none()
+    }
+}
+
+impl FromPyObject<'_> for UrlHostParts {
+    fn extract(ob: &'_ PyAny) -> PyResult<Self> {
+        let py = ob.py();
+        let dict = ob.downcast::<PyDict>()?;
+        Ok(UrlHostParts {
+            username: dict.get_as(intern!(py, "username"))?,
+            password: dict.get_as(intern!(py, "password"))?,
+            host: dict.get_as(intern!(py, "host"))?,
+            port: dict.get_as(intern!(py, "port"))?,
+        })
+    }
+}
+
+impl fmt::Display for UrlHostParts {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        let _ = match self {
+            UrlHostParts {
+                username: Some(username),
+                password: None,
+                ..
+            } => write!(f, "{username}"),
+            UrlHostParts {
+                username: None,
+                password: Some(password),
+                ..
+            } => write!(f, ":{password}"),
+            UrlHostParts {
+                username: Some(username),
+                password: Some(password),
+                ..
+            } => write!(f, "{username}:{password}"),
+            UrlHostParts {
+                username: None,
+                password: None,
+                ..
+            } => Ok(()),
+        };
+        if self.host.is_some() {
+            write!(f, "@{}", self.host.as_ref().unwrap())?;
+        }
+        if self.port.is_some() {
+            write!(f, ":{}", self.port.as_ref().unwrap())?;
+        }
```

Comment:
I'll push one tiny style fix here, and also move the `@` from the `host` branch into the cases where username / password is set.
```suggestion
        match (&self.username, &self.password) {
            (Some(username), None) => write!(f, "{username}@")?,
            (None, Some(password)) => write!(f, ":{password}@")?,
            (Some(username), Some(password)) => write!(f, "{username}:{password}@")?,
            (None, None) => {},
        };
        if let Some(host) = &self.host {
            write!(f, "{host}")?;
        }
        if let Some(port) = self.port {
            write!(f, ":{port}")?;
        }
```

Reply:
Aaaah makes sense. I was having issues trying to get it to use the fields so thought the struct match might be _the way_ to do it, but great to know this for the future.